src/Analysis/Engine/Impl/Infrastructure/TextBuilder.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.PythonTools.Analysis.Infrastructure {
+    internal class TextBuilder {
```

Comment:
@jakebailey, what is the benefit of having this class instead of common `StringBuilder`? `SoftAppendSpace` is specific for `LineFormatter`, and while there is no built-in method to find the last non-whitespace character, it is very simple to implement.

Reply:
This class is pulled from the vscode-python extension. I need this sort of class in order to be able to insert spaces without duplicating them (unless that is needed). I originally avoided StringBuilder because I thought I would need to invoke ToString repeatedly to figure out how the builder ends, but looking at StringBuilderExtensions, I guess you can just index into the builder and let it figure it out.

If you'd prefer, I could probably write an extension that does something along the lines of ensuring there's some number of spaces at the end of a builder (and just trim it at the end) and remove the TextBuilder.

Reply:
I've written a single extension method for StringBuilder that replaces TextBuilder entirely:
```csharp
        public static StringBuilder EnsureEndsWithSpace(this StringBuilder sb, int count = 1, bool allowLeading = false) {
            if (sb.Length == 0 && !allowLeading) {
                return sb;
            }

            for (var i = sb.Length - 1; i >= 0; i--) {
                if (!char.IsWhiteSpace(sb[i])) {
                    break;
                }
                count--;
            }

            if (count > 0) {
                sb.Append(new string(' ', count));
            }

            return sb;
        }
```
If there's no objection I can remove TextBuilder and just use this.

Reply:
OK. In RTVS and other places I used it in it was handling quite a bit more than just trailing ws - such as type of indentation, conversion of spaces to tabs and vice versa, etc. SB would be very limiting there.